### PR TITLE
Do some optimization according to the review comments #3400

### PR DESCRIPTION
**Pull request by majorcool**
**Fixes issue:**
Do some optimization according to the review comments

**Changes:**
Binary search in javascript

### CODE REVIEW

File: code/search/src/binary_search/binary_search.js [hunk 1]
```diff
@@ -1,18 +1,45 @@
...
-            lo = mid + 1;
-        } else if (element > key) {
-            hi = mid - 1;
-        } else {
-            return mid;
-        }
+// implementation by looping
+function binarySearchLooping(array, key) {
+  let lo = 0,
+    hi = array.length - 1;
```
Comment (arnavb) [inline code/search/src/binary_search/binary_search.js]: Add a newline below.

File: code/search/src/binary_search/binary_search.js [hunk 2]
```diff
@@ -1,18 +1,45 @@
...
+    let mid = Math.floor((lo + hi) / 2, 10);
+    let element = array[mid];
+    if (element < key) {
+      lo = mid + 1;
+    } else if (element > key) {
+      hi = mid - 1;
+    } else {
+      return mid;
     }
+  }
```
Comment (arnavb) [inline code/search/src/binary_search/binary_search.js]: Add a newline below.

File: code/search/src/binary_search/binary_search.js [hunk 3]
```diff
@@ -1,18 +1,45 @@
...
+/**
+ *
+ * @param {* number[]} arr - the sorted array to be searched in
+ * @param {* number} value - the value to be searched
+ * @param {* number} low - the start index of the search range
+ * @param {* number} high - the end index of the search range
+ */
+function binarySearchByRecursion(arr, value, low = 0, high = arr.length - 1) {
+  const mid = Math.floor((start + end) / 2);
+  if (end === start + 1 && arr[mid] !== value && arr[start] !== value) {
```
Comment (arnavb) [inline code/search/src/binary_search/binary_search.js]: Add newline above.

File: code/search/src/binary_search/binary_search.js [hunk 4]
```diff
@@ -1,18 +1,45 @@
...
     return -1;
+  }
+  if (arr[start] === value) {
+    return start;
+  } else if (arr[end] === value) {
+    return end;
+  } else if (arr[mid] === value) {
+    return mid;
+  } else {
+    if (arr[mid] > value) {
```
Comment (arnavb) [inline code/search/src/binary_search/binary_search.js]: Use ternary here:

```Javascript
return arr[mid] > value ? 
  binarySearchByRecursion(arr, value, start, mid) : 
  binarySearchByRecursion(arr, value, mid, end);
```

File: code/search/src/interpolation_search/interpolation_search.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+function interpolationSearchRecursion(
+  arr,
+  value,
+  low = 0,
+  high = arr.length - 1
+) {
+  if (arr.length === 0 || (arr.length === 1 && arr[0] !== value)) {
+    return -1;
+  }
+  let mid =
```
Comment (arnavb) [inline code/search/src/interpolation_search/interpolation_search.js]: Move `low +` to the line above.

File: code/search/src/interpolation_search/interpolation_search.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  } else {
+    return interpolationSearchRecursion(arr, value, mid + 1, high);
+  }
+}
+
+// implementation with recursion
+function interpolationSearchLooping(arr, value) {
+  let low = 0,
+    high = arr.length - 1;
+  while (low < high) {
```
Comment (arnavb) [inline code/search/src/interpolation_search/interpolation_search.js]: Add newline above.

File: code/sorting/src/gnome_sort/gnome_sort.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+function gnomeSort(arr) {
+  let index = 0,
+    len = arr.length;
+  while (index <= len) {
```
Comment (arnavb) [inline code/sorting/src/gnome_sort/gnome_sort.js]: Add newline above. Also, I find `len` a useless variable, so I would remove it.

File: code/sorting/src/gnome_sort/gnome_sort.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    if (index > 0 && arr[index - 1] > arr[index]) {
+      let tmp = arr[index];
+      arr[index] = arr[index - 1];
+      arr[index - 1] = tmp;
+      index--;
```
Comment (arnavb) [inline code/sorting/src/gnome_sort/gnome_sort.js]: Preincrement/Predecrement whenever possible.

**Comment by majorcool** [comment]
Thanks for all the suggestions. I have finished prettifying and simplifying those code you mentioned above. Please check it out when you are free.

**Comment by AdiChat** [comment]
Looks great 😍 

Hope you are enjoying your journey 🚆  with **OpenGenus**

**Pull request merged**